Android wear crash on startup

logs

E/AndroidRuntime( 9820): java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
E/AndroidRuntime( 9820): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
E/AndroidRuntime( 9820): Caused by: java.lang.reflect.InvocationTargetException
E/AndroidRuntime( 9820): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 9820): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime( 9820): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
E/AndroidRuntime( 9820): 	... 1 more
E/AndroidRuntime( 9820): Caused by: md52ce486a14f4bcd95899665e9d932190b.JavaProxyThrowable: System.FormatException: String was not recognized as a valid DateTime.
E/AndroidRuntime( 9820):   at System.DateTimeParse.Parse (System.String s, System.Globalization.DateTimeFormatInfo dtfi, DateTimeStyles styles) <0xa98bdba0 + 0x00094> in <filename unknown>:0
E/AndroidRuntime( 9820):   at System.DateTime.Parse (System.String s) <0xa98bae20 + 0x00027> in <filename unknown>:0
E/AndroidRuntime( 9820):   at Toggl.Chandler.Common.Get[T] (Android.Gms.Wearable.DataMap dataMap, System.String propertyName, Toggl.Chandler.T defaultValue) <0xa98baba0 + 0x00163> in <filename unknown>:0
E/AndroidRuntime( 9820):   at Toggl.Chandler.SimpleTimeEntryData..ctor (Android.Gms.Wearable.DataMap map) <0xa98b6550 + 0x001b3> in <filename unknown>:0
E/AndroidRuntime( 9820):   at Toggl.Chandler.UI.Activities.MainActivity.OnDataChanged (IDataItem dataItem) <0xa98b5110 + 0x00117> in <filename unknown>:0
E/AndroidRuntime( 9820):   at Toggl.Chandler.UI.Activities.MainActivity.OnDataChanged (Android.Gms.Wearable.DataEventBuffer dataEvents) <0xa98b3658 + 0x00193> in <filename unknown>:0
E/AndroidRuntime( 9820):   at Android.Gms.Wearable.IDataApiDataListenerInvoker.n_OnDataChanged_Lcom_google_android_gms_wearable_DataEventBuffer_ (IntPtr jnienv, IntPtr native__this, IntPtr native_dataEvents) <0xa98b33a8 + 0x00063> in <filename unknown>:0
E/AndroidRuntime( 9820):   at (wrapper dynamic-method) System.Object:56860e3b-8848-498e-971a-df168bc43803 (intptr,intptr,intptr)
E/AndroidRuntime( 9820): 	at md532180f774cdaabd03ff1344b502b0625.MainActivity.n_onDataChanged(Native Method)
E/AndroidRuntime( 9820): 	at md532180f774cdaabd03ff1344b502b0625.MainActivity.onDataChanged(MainActivity.java:61)
E/AndroidRuntime( 9820): 	at com.google.android.gms.wearable.internal.zzbp$3.zza(Unknown Source)
E/AndroidRuntime( 9820): 	at com.google.android.gms.wearable.internal.zzbp$3.zzq(Unknown Source)
E/AndroidRuntime( 9820): 	at com.google.android.gms.internal.zzlm.zzb(Unknown Source)
E/AndroidRuntime( 9820): 	at com.google.android.gms.internal.zzlm$zza.handleMessage(Unknown Source)
E/AndroidRuntime( 9820): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime( 9820): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime( 9820): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
E/AndroidRuntime( 9820): 	... 4 more
W/ActivityManager(  582):   Force finishing activity 1 com.toggl.timer/md532180f774cdaabd03ff1344b502b0625.MainActivity
W/ActivityManager(  582): Activity pause timeout for ActivityRecord{3b428f01 u0 com.toggl.timer/md532180f774cdaabd03ff1344b502b0625.MainActivity t22 f}
D/AmbientService( 1268): onResumed: AmbientActivity[component=ComponentInfo{com.google.android.wearable.app/com.google.android.clockwork.home.HomeActivity}, callbacks=@365e9bc4, running=false, ambient=false] Binder_1 79
I/AudioFlinger(  197): AudioFlinger's thread 0xb583e008 ready to run
W/bt-btif ( 1337): dm_pm_timer expires
W/bt-btif ( 1337): dm_pm_timer expires 0

--- a/Chandler/Common.cs
+++ b/Chandler/Common.cs
@@ -46,7 +46,9 @@ namespace Toggl.Chandler
                     return (T) (object)Guid.Parse (s);
                 } else if (t == typeof (DateTime)) {
                     var s = dataMap.GetString (key);
-                    return (T) (object)DateTime.Parse (s);
+                    DateTime dt = new DateTime();
+                    DateTime.TryParse (s, out dt);
+                    return (T) (object) dt;
                 } else {
                     throw new NotSupportedException (typeof (T).Name);
                 }

--- a/Chandler/UI/Activities/MainActivity.cs
+++ b/Chandler/UI/Activities/MainActivity.cs
@@ -24,7 +24,7 @@ namespace Toggl.Chandler.UI.Activities
         private GoogleApiClient googleApiClient;
         private PagesAdapter adapter;
         private List<SimpleTimeEntryData> timeEntries = new List<SimpleTimeEntryData> ();
-        private const int RebindTime = 2000;
+        private const int RebindTime = 4000;
 
         protected override void OnCreate (Bundle savedInstanceState)
         {

--- a/Joey/Wear/Common.cs
+++ b/Joey/Wear/Common.cs
@@ -46,7 +46,9 @@ namespace Toggl.Joey.Wear
                     return (T) (object)Guid.Parse (s);
                 } else if (t == typeof (DateTime)) {
                     var s = dataMap.GetString (key);
-                    return (T) (object)DateTime.Parse (s);
+                    DateTime dt = new DateTime();
+                    DateTime.TryParse (s, out dt);
+                    return (T) (object) dt;
                 } else {
                     throw new NotSupportedException (typeof (T).Name);
                 }

--- a/Joey/Wear/WearDataService.cs
+++ b/Joey/Wear/WearDataService.cs
@@ -78,6 +78,7 @@ namespace Toggl.Joey.Wear
 
         public async void OnConnected (Android.OS.Bundle connectionHint)
         {
+            Log.Info ("WearIntegration", "Connected");
             await UpdateSharedTimeEntryList();
         }
 
@@ -107,12 +108,14 @@ namespace Toggl.Joey.Wear
             try {
                 Log.Info ("WearIntegration", "Received Message");
 
+                googleApiClient.Connect();
                 if (!googleApiClient.IsConnected) {
-                    googleApiClient.Connect();
+                    Log.Info ("WearIntegration", "Connecting");
                 }
 
                 var authManager = ServiceContainer.Resolve<AuthManager> ();
                 if (!authManager.IsAuthenticated) {
+                    Log.Info ("WearIntegration", "Is not authenticated");
                     NotifyNotLoggedIn();
                     return;
                 }
@@ -130,6 +133,7 @@ namespace Toggl.Joey.Wear
                         await StartEntry (guid);
                         await UpdateSharedTimeEntryList ();
                     } else if (path == Common.RequestSyncPath) {
+                        Log.Info ("WearIntegration", "Sending sync data!");
 
                         await UpdateSharedTimeEntryList ();
                     } else if (path == Common.OpenHandheldPath) {
@@ -183,7 +187,28 @@ namespace Toggl.Joey.Wear
                 currentDataMap.Add (entry.DataMap);
             }
             mapReq.DataMap.PutDataMapArrayList (Common.TimeEntryListKey, currentDataMap);
-            await WearableClass.DataApi.PutDataItem (googleApiClient, mapReq.AsPutDataRequest ());
+            await SendData (mapReq);
+        }
+
+        private Task SendData (PutDataMapRequest mapReq)
+        {
+            return Task.Run (() => {
+                foreach (var node in clientNodes) {
+                    WearableClass.DataApi.PutDataItem (googleApiClient, mapReq.AsPutDataRequest());
+                }
+            });
+        }
+
+        private IList<INode> clientNodes
+        {
+            get {
+                return WearableClass
+                       .NodeApi
+                       .GetConnectedNodes (googleApiClient)
+                       .Await ()
+                       .JavaCast<INodeApiGetConnectedNodesResult> ()
+                       .Nodes;
+            }
         }
 
         public static void LOGD (string tag, string message)
@@ -194,4 +219,3 @@ namespace Toggl.Joey.Wear
         }
     }
 }
-

